cql3: expr: make it possible to evaluate expr::binary_operator

cvybhu opened the pull request:
As a part of CQL rewrite we want to be able to perform filtering by calling `evaluate()` on an expression and checking if it evaluates to `true`. Currently trying to do that for a binary operator would result in an error.

Right now checking if a binary operation like `col1 = 123` is true is done using `is_satisfied_by`, which is able to check if a binary operation evaluates to true for a small set of predefined cases.

Eventually once the grammar is relaxed we will be able to write expressions like: `(col1 < col2) = (1 > ?)`, which doesn't fit with what `is_satisfied_by` is supposed to do.
Additionally expressions like `1 = NULL` should evaluate to `NULL`, not `true` or `false`. `is_satsified_by` is not able to express that properly.

The proper way to go is implementing `evaluate(binary_operator)`, which takes a binary operation and returns what the result of it would be.

Implementing `prepare_expression` for `binary_operator` requires us to be able to evaluate it first. In the next PR I will add support for `prepare_expression`.

nyh commented:
Hi @cvybhu I'm a bit hazy about the grand plan, so I have a question which might be stupid. Basically my question stems from wondering whether after this patch the `evaluate()` you're implementing here will be used in the query code, and how.

Definitely it makes sense to support arbitrary expressions like `(col1 < col2) = (1 > ?)` and `evaluate()` could do that.  This could be useful for `ALLOW FILTERING`. However, a big part of Scylla's query implementation is figuring out what can be done efficiently using partition keys, clustering key slices, and indexes - so for example `col1 = 1` can be efficient when `col1` is the partition key, but `(col1 < col2) = (1 > ?)` cannot be implemented without a full scan. If I understand correctly, `evaluate()` will not be useful for this, and will only be used for `ALLOW FILTERING` - but if I understand correctly, it is not yet used. 
Is my understanding correct?

avikivity commented:
One of the goals is to make the query planner (the part that decides which parts of the WHERE clause are used to construct read_command and which are used for filtering, and whether an index is used) more explicit and easier to understand.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -1728,10 +1728,60 @@ std::optional<bool> get_bool_value(const constant& constant_val) {
+        }
+        default: {
+            throw exceptions::unsupported_operation_exception(format("Unhandled binary_operator: {}", binop));
+        }
```
Ah, I see the following patch handles it.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -516,105 +516,30 @@ value_set intersection(value_set a, value_set b, const abstract_type* type) {
-        }, opr.lhs);
+    if (is<token>(opr.lhs)) {
+        // The RHS value was already used to ensure we fetch only rows in the specified
+        // token range. It is impossible for any fetched row not to match now.
```
Note, this won't be true one day. We should compute the token and apply the binary operator to it (and if the query planner converted that subexpression to read_command, remove it from filtering).

oh wait, the following comment says exactly that :facepalm:

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -233,21 +233,6 @@ bool equal(const expression& lhs, const expression& rhs, const evaluation_inputs
-        return equal(lhs, b, inputs);
-    });
-}
-
```
:+1:

avikivity reviewed (approved):
Very good, especially liked the removed functions.

nyh commented on cql3/expr/expression.cc:
```diff
@@ -216,21 +216,95 @@ get_value(const subscript& s, const evaluation_inputs& inputs) {
+class bool_or_null {
+    std::optional<bool> value;
+public:
+    bool_or_null(bool val) : value(std::move(val)) {}
```
Nitpick: no need to std::move a primitive bool type.

nyh commented on cql3/expr/expression.cc:
```diff
@@ -216,21 +216,95 @@ get_value(const subscript& s, const evaluation_inputs& inputs) {
+    return type_of(lhs)->equal(managed_bytes_view(lhs_bytes), managed_bytes_view(*rhs_bytes));
+}
+
+static std::optional<std::pair<managed_bytes, managed_bytes>> evaluate_binop_sides(const expression& lhs,
```
Oh, so it seems here you're trying to do what I asked above - have one place int he code that evaluates both RHS and LHS and checks for UNSET and so on, instead of repeating it in multiple different operator. Good.

nyh commented on cql3/expr/expression.cc:
```diff
@@ -216,21 +216,95 @@ get_value(const subscript& s, const evaluation_inputs& inputs) {
+            format("UNSET_VALUE found on right-hand side of a binary operator with operation {}", op));
+    }
+    if (lhs_value.is_null() || rhs_value.is_null()) {
+        return std::nullopt;
```
Wait, what if the operation is IS NOT NULL? Don't we need to return TRUE for that, not NULL?

nyh commented on cql3/expr/expression.cc:
```diff
@@ -343,43 +343,37 @@ bool_or_null limits(const expression& lhs, oper_t op, const expression& rhs, con
+        return exists_in(data_map | transformed([](const entry& e) { return e.second; }));
     } else {
-        return false;
+        on_internal_error(expr_logger, "unsupported collection type in a CONTAINS expression");
```
I hope we have an earlier check that lhs is indeed a collection, and a malicious request cannot trigger this case?

cvybhu replied:
`prepare_expression` should perform all necessary type checks. `evaluate()` usually assumes that the types are correct, but there are some type checks sprinkled here and there.

cvybhu replied:
`prepare_expression` should perform all type checks. `evaluate()` usually assumes that the types were checked and are correct, although there are some type checks sprinkled here and there.

nyh reviewed (approved):
Looks good. I left quite a few comments, but I think nothing is really blocking and this is code is being incrementally improved anyway.

I'm happy about the unit tests, but eventually when the new evaluate() code actually gets used in CQL (e.g., ALLOW FILTERING), I'd like to see it all tested in a CQL test and not just in C++. The problem is that it's not always clear how exactly the CQL expression is translated to all these C++ concepts, and the error cases are especially unclear because we have various layers of validity checks that may cause the final result returned by an expression to be different than the result returned by one specific C++ function.
Testing in CQL will also allow us to:
1. Run the same test against older Scylla and a post-rewrite Scylla and check we didn't unintentionally break any expressions that used to work.
2. Run the same test against Cassandra, and see that if our result differs from Cassandra, then this difference is intentional and Scylla's behavior is better. I think we should document all these differences in a new document on how Scylla's expressions differ from Cassandra's (or older Scylla versions).

avikivity commented:
> Looks good. I left quite a few comments, but I think nothing is really blocking and this is code is being incrementally improved anyway.
> 
> I'm happy about the unit tests, but eventually when the new evaluate() code actually gets used in CQL (e.g., ALLOW FILTERING), I'd like to see it all tested in a CQL test and not just in C++.

It's already tested in cql, no? All those filtering expressions should already have tests.

> The problem is that it's not always clear how exactly the CQL expression is translated to all these C++ concepts, and the error cases are especially unclear because we have various layers of validity checks that may cause the final result returned by an expression to be different than the result returned by one specific C++ function. Testing in CQL will also allow us to:
> 
> 1. Run the same test against older Scylla and a post-rewrite Scylla and check we didn't unintentionally break any expressions that used to work.
> 2. Run the same test against Cassandra, and see that if our result differs from Cassandra, then this difference is intentional and Scylla's behavior is better. I think we should document all these differences in a new document on how Scylla's expressions differ from Cassandra's (or older Scylla versions).

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3108/

nyh commented:
> > I'm happy about the unit tests, but eventually when the new evaluate() code actually gets used in CQL (e.g., ALLOW FILTERING), I'd like to see it all tested in a CQL test and not just in C++.
> 
> It's already tested in cql, no? All those filtering expressions should already have tests.

I don't know if we do. Which tests?

Also, during this work Jan is discovering interesting edge cases that he had to specially handle in code. Do these specific edge cases have tests? Possibly not. (But again, I'm not sure, maybe Dejan wrote in the past some tests in this area that I'm not aware of)?

Again, I agree that the C++ unit tests are much better than nothing. I'm just worried that because the real CQL code includes multiple layers - parser, query planner, etc., we're not even sure that a specific function like evaluate() will be called when it should. As a made-up example, consider that maybe this patch fixes an error in a filter "a = null" when a is a regular column (and allow filtering). But for p = null, for partition key, different code is reached. Sometimes its not obvious which code is reached.

And end-to-end CQL test is also a better regression test: tests in test/cql-pytest and even tesf/cql can be repeated before and after a rewrite, and confirm that the rewrite doesn't break anything. The C++ unit tests can't do this - the new unit tests that Jan wrote now can't be run on older Scylla. If the rewrite mistakenly changed the behavior of some request, these tests won't be able to discover that.

cvybhu commented on cql3/expr/expression.cc:
```diff
@@ -1728,10 +1728,60 @@ std::optional<bool> get_bool_value(const constant& constant_val) {
+    bool binop_result;
+
+    switch (binop.op) {
+        case oper_t::EQ: {
```
:+1: removed

cvybhu commented on cql3/expr/expression.cc:
```diff
@@ -516,105 +516,30 @@ value_set intersection(value_set a, value_set b, const abstract_type* type) {
+    if (is<token>(opr.lhs)) {
+        // The RHS value was already used to ensure we fetch only rows in the specified
+        // token range. It is impossible for any fetched row not to match now.
+        // When token restrictions are present we forbid all other restrictions on partition key.
```
Added a test in [84501851eb709c649fd478008da5b4c0b076ec43](https://github.com/scylladb/scylladb/pull/12052/commits/84501851eb709c649fd478008da5b4c0b076ec43)

cvybhu commented on cql3/expr/expression.cc:
```diff
@@ -216,21 +216,95 @@ get_value(const subscript& s, const evaluation_inputs& inputs) {
 }
 
+// This class represents a value that can be one of three things:
+// false, true or null.
```
Well there is:
*  `raw_value`, which can represent `true`, `false`, `empty`, `null` and `unset`. That is two more cases that have to be checked everywhere.
* `data_value`, which is quite complicated and AFAIU requires an allocation for each value, which would slow down filtering.

`bool_or_null` is readable and easy to understand. I personally like it, but maybe `data_value` could also be used here.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -1728,10 +1728,60 @@ std::optional<bool> get_bool_value(const constant& constant_val) {
+        }
+        case oper_t::CONTAINS: {
+            cql3::raw_value val = evaluate(binop.rhs, inputs);
+            binop_result = contains(binop.lhs, val.view(), inputs);
```
History?

cvybhu commented:
> > > I'm happy about the unit tests, but eventually when the new evaluate() code actually gets used in CQL (e.g., ALLOW FILTERING), I'd like to see it all tested in a CQL test and not just in C++.

I agree that it would be great to also have some CQL tests. In these unit tests I always have `constant` on both sides of a `binary_operator`, but there might be other stuff there, like `tuple_constructors` and so on.
There's a bunch of ifs that these unit tests don't hit.

In unit tests I can test a reasonably exhaustive set of cases, because I know that both sides will be evaluated and have a few possible values. In CQL there are many more cases, various combinations of partition restrictions, clustering etc...
My fingers already hurt thinking about enumerating all possible edge cases where partition key is UNSET, but clustering is EMPTY and there is no token restrictions....
I hope that automated CQL tests could be done in a way that tests all those combinations for us.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3131/

avikivity commented:
Since is_satisfied_by was reimplemented in terms of evaluate(), I think we have good coverage. Once we relax the grammar we can add some additional cql tests, but as of now our coverage only increased relative to pre-patch.

cvybhu commented:
Hmm this PR fixes #10358 in the sense that now we always throw an error when there is a comparison with UNSET_VALUE.
The tests xfailed because of #10358 still don't pass because the error messages differ.

I think the easiest way to fix it would be to modify the tests so that they expect an error message containing `unset value` instead of the exact one that Cassandra outputs.

Implementing the exact same messages as in Cassandra would require some acrobatics, the message is even different if a column has an index (!).

https://github.com/scylladb/scylladb/blob/996eac9569fba1b08ae15e2533a158fd0d23b214/test/cql-pytest/cassandra_tests/validation/operations/select_single_column_relation_test.py#L491-L516

avikivity commented:
Agree with relaxing the test.

As you may remember, I want to move all unset processing to the prepare phase and completely exclude it from evaluation. So it's not worth spending effort on this misfeature.

cvybhu commented:
v2:
* Modified error messages for `unset value` in invalid places to match the tests
* Modified the tests to accept all error messages with `unset value` inside of them instead of exactly the messages that Cassandra produces
* Enabled some tests that were xfailed, but started passing thanks to this PR

cvybhu commented:
It turns out that this PR doesn't full fix https://github.com/scylladb/scylladb/issues/10358.
The new checks for `unset value` only get executed when filtering, all the non-filtering code still accepts `unset value` even though it shouldn't.

Restrictions like `WHERE pk = ?` will evaluate the rhs, and happily accept that this is an `unset_value`.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3171/

nyh commented:
> In unit tests I can test a reasonably exhaustive set of cases, because I know that both sides will be evaluated and have a few possible values. In CQL there are many more cases, various combinations of partition restrictions, clustering etc... My fingers already hurt thinking about enumerating all possible edge cases where partition key is UNSET, but clustering is EMPTY and there is no token restrictions.... I hope that automated CQL tests could be done in a way that tests all those combinations for us.

I think "automated" (usually this means "random") tests aren't necessarily bad, but they are overrated. The whole point of the cql-pytest framework, for example, was to make it easy to write explicit tests - so it's not really harder to "enumerate" many different cases in Python than in C++ code - often it's the opposite, the Python code is easier to write.

But my thinking is that if you write tests before **and** during writing the code, you know exactly which corner cases you handled in code and can write tests for them. It is true that you may miss some corner cases you didn't consider to test, but what you do get is quality **regression testing**: If a year from now you or somebody else refactors the code, they cannot break any of the use cases you tested. 

This is in fact exactly what happened to you here: with https://github.com/scylladb/scylladb/issues/10358 - you are refactoring (or in this case, writing from scratch) code for handling UNSET_VALUE in SELECT, but your refactoring forgot a whole class of cases (selects using a partition or index, not a filter) and the regression test that we have (that the Cassandra developers wrote for us, and I translated) caught that you forgot to handle that case.

This - worry-free refactoring - is something that low-level C++ unit tests often cannot get you. When the low-level code is refactored, often the test needs to be thrown away.

But I guess it can't hurt to have both kinds of tests :-)

cvybhu commented:
The CI failed twice with an error that doesn't seem related to the PR:
```
20:36:17  ERROR: Could not get unified-pack-url-x86_64 from relocatable metadata file
20:36:17  Finished: FAILURE
```

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3197/

avikivity reviewed (approved):
Looks good, will wait for @nyh to confirm his wants are addressed.
